### PR TITLE
feat: add NumberGlyph visualizer

### PR DESCRIPTION
**Pull request by liammulh**
By submitting this PR, I am indicating to the Numberscope maintainers that I have read and understood the [contributing guidelines](https://numberscope.colorado.edu/doc/CONTRIBUTING/) and that this PR follows those guidelines to the best of my knowledge. I have also read the [pull request checklist](https://numberscope.colorado.edu/doc/doc/pull-request-checklist/) and followed the instructions therein.

<hr/>

This is @LeongHouHeng's sequence color visualizer PR. We are continuing it based on a branch in this repository. The original PR, which was submitted from a fork is https://github.com/numberscope/frontscope/pull/191.

### CODE REVIEW

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:40]: When I submitted my Chaos Visualizer for a PR, @gwhitney encouraged me to use the p5.Vector class which has attributes x and y. See the chaos visualizer code [here](https://github.com/numberscope/frontscope/blob/bb157252d6ce8dfb603bbb0860531daee34a3053/src/visualizers/VisualizerChaos.ts#L236). That might be nice for some of your x and y coordinate variables.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:110]: I'm finding the following behaviour: When I initially create a bundle, it draws the small version with coloured balls. Then when I click "Draw" and go into the main canvas, it is black/blank. The left key works to show/hide the legend. But no circles appear until I hit the right key for "refresh". So maybe it needs to have an automatic refresh somehow when moving from bundle card to main canvas and back? Would a change of this.firstDraw bool go into setup()? I'm not exactly sure.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:14]: Maybe you could explain that the first circle is associated to the first term of the sequence, and the second circle to the second number in the sequence, etc. You could also say that the fuzziness of the ball is a visual indicator: a brighter, less fuzzy ball means that term is bigger.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:129]: What about negative terms? Right now it draws black. I suggest you just take the absolute values of terms, so that the sequences will converge. If you do this, you should also document this behaviour in the documentation above.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:132]: The visualizer will hang pretty quickly on very large numbers (of which there are plenty in the OEIS). You might want to have a cut-off for a, where if you see a number bigger than that cutoff, you just "fake it" by drawing a pretty non-fuzzy bright circle (I'm assuming that as the numbers get very big, the eye won't detect the "fake" here; maybe I'm wrong?).

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:12]: Another thought: since a lot of sequences (including just the natural numbers) get to a bright full ball fairly quickly (in other words, all balls look pretty bright), maybe we could somehow "tune down" the brightness (e.g. run the given algorithm on a/10 or on log(a)) until we can see more a visual difference between, say natural numbers and squares?

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:100]: Another thing that might be nice is to be able to toggle the labels (the numbers 1/a^n) on and off with a key. That way, when we turn them off, we can have a nice pure visual. You could even have them toggle on/off along with the legend (left key). So when legend is on, numbers show. When legend is off, numbers are gone.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:124]: I've noticed what I think is a bug. Try the visualizer with the sequence from formula n, n+1, n+2, n+3 etc. The circles should just be "shifting left" between each of these. But the first circle is not fuzzy, it looks like a single disk, e.g. n+2. So something funny is happening with the first circle. (I've added this comment near where the disc is calculated, but I don't know what the bug is, I may be in the wrong place in the code.)

**Review by katestange — CHANGES_REQUESTED** [review]
This is transferring the review comments from the old branch.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts:8]: Do the colors have anything to do with the sequence itself? Or are they simply indicating which divisor sequence a particular disk is using, i.e. are the same colors regardless of the sequence? If they are always the same, then maybe the colors are not really "part of" the visualization in some sense, i.e., the colors aren't telling you anything about the sequence you're analyzing, and then shouldn't be part of the name? So maybe just "Growth Charts"?

Comment (katestange) [reply src/visualizers/SeqColor.ts:8]: The colours are the prime factorizations of the terms.  I think basically the colour represents prime factorization and the brightness represents size in a way.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:108]: This bug needs resolving.  I gave it a go, however, it looks like it requires `keyPressed()` from p5.js, and this isn't working; maybe similar to #120 ?

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:296]: `isPrime()` and `primeFactors()` should use `sequence.getFactors()`

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:110]: I can no longer reproduce this behaviour.

Comment (katestange) [reply src/visualizers/SeqColor.ts:110]: Actually, with A095102, for example, the refresh (right arrow) does alter the image.  So there still is some version of this bug going on.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: done; check it out

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: changed; I went for "Primes and Sizes"

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: done; you can take a look (repeats code from PR #230 (isPrime) but this is temporary I suppose)

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:9]: The visualizer needs to be fully documented.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:100]: This is actually already implemented, so this comment is no longer relevant.  Resolving.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:217]: In a recent commit, I made it so the user can decide the growth function by formula, which should alleviate this (and besides, now it would be very difficult to do this).  So I would like to simply resolve this conversation -- no action suggested.  Sound ok?

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:15]: This is effectively resolved by the recent commit where I added a formula box so you can control the growth by formula.  So I think this is resolved.  If you agree, please resolve.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:149]: this is odd, because it takes the prime values of the sequence, not prime divisors?  is that what we want?  those are the only primes which contribute to the colouring of the sequence, as the code stands.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:228]: This took a lot to track down!  It was because we needed to call `this.sketch.noStroke()` -- it was drawing the circle but with a black border that overwrote the colour of the last, larger circle each time.  So it was there, but very hard to see.  Somewhere else `noStroke()` gets set before it draws the next circle.  This is fixed in the most recent commit.  Please check it out.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:231]: fixed in a recent commit (except not yet documented).  now it takes the absolute value of any growth function

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:189]: This seems to be fixed with the other bug fix.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:285]: When we toggle the labels and the legend-box, sometimes the redraw doesn't clear the background, so we end up with circles over top of the white label box.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: fixed; take a look

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:299]: For discussion: should we try to make this visualizer work with bigint?  It allows for arbitrary formulas in terms of the input values based on @gwhitney 's code, so I don't know how feasible that is.

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:558]: Still to do on documentation:  add more examples.  Note that once the grid vis is in, the latex here will work also.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: This should be resolved with the new docs.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts:10]: mostly done now; still needs maybe another interesting example or two

File: src/visualizers/SeqColor.ts
Comment (katestange) [inline src/visualizers/SeqColor.ts:43]: still a bug with a095102, which seems to miss the top right disc when using the default settings

Comment (katestange) [reply src/visualizers/SeqColor.ts:43]: fixed now

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: an example of semiprimes added, it's probably enough for now.

**Comment by liammulh** [comment]
Oh no! I just realized that when I transferred the old comments a couple weeks ago I didn't click the "Finish your review" button, and I guess maybe that made it so that other people couldn't see that I had transferred the comments from the old PR over? It looks like @katestange also transferred the comments. I'll try to delete any comments that are duplicates.

**Comment by liammulh** [comment]
Okay, I think I got rid of all the duplicate comments.

**Comment by gwhitney** [comment]
This PR uses a non-conventional name for the visualizer: SeqColor.ts, rather than VisualizerPrimesSizes.ts. Either it the implementation of this visualizer should have its filename renamed in the "standard" way, or we should rename all of the others (as in just Chaos.ts, Turtle.ts, Differences.ts, etc.). There is definitely an argument for the latter, as having "Visualizer" in the name of every source file in the "src/visualizer" directory is kind of redundant. We should discuss tomorrow which way to go and make it so.

**Comment by gwhitney** [comment]
The image files for the documentation have not yet been included in any commit in this PR. Could whoever has them please commit and push them? Thanks.

**Comment by gwhitney** [comment]
Checklist of items that need to be taken care of in this PR, from group discussion today. If you are the one who verifies that one of these things has been done by a new commit to this PR, please check the item off. Note that this is not necessarily exhaustive, there may well still be additional items that need resolution in the conversations above.

- [ ] Rename to NumberGlyph visualizer
- [ ] make the parameter dialog initially have a single 'customize glyph generation' boolean parameter and the other other parameters are only shown when that is checked
- [ ] Commit the images needed in the documentation
- [ ] Rebase on PRs solving #120 and/or #241 when they are merged
- [ ] Change the default glyph-generating function to something that looks better, including defaulting growth rings to true
- [ ] Make code check for safety of conversion to number (note that sequence `2^n%17` was not coming out periodic, probably because of this bug, although it may not be something that can be handled in this way because the problem there is _inside_ mathjs; the inputs and outputs are OK numbers, but they will eventually be wrong because of a internal overflow in mathjs).
- [ ] Warn in docs about needing to fit into number (for now)
- [ ] Deal with integers with no prime factors so that they are not just blank black spaces (maybe use grey?)
- [ ] Once #120 is fixed, improve situation so that toggling the legend doesn't "flicker"

**Comment by katestange** [comment]
My most recent commit attempts to handle possible bigint errors but it doesn't do a great job... suggestions appreciated.

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: we decided to document its failings; see [comment here](https://github.com/numberscope/frontscope/pull/231#issuecomment-1377967936)

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: I'm quite happy with this now; ready for review

File: src/visualizers/SeqColor.ts
Comment (katestange) [reply src/visualizers/SeqColor.ts]: See  [comment here](https://github.com/numberscope/frontscope/pull/231#issuecomment-1377967936) -- it was renamed to glyphs.

**Comment by katestange** [comment]
I believe everything is dealt with here except for the issue that is blocked by issue #120 .  Please take a look and resolve the open conversations or give feedback.

**Review by katestange — APPROVED** [review]

**Comment by gwhitney** [comment]
I took the liberty of fitting the square of glyphs to the canvas, and I found a way in this case of avoiding the `{} as p5.Vector` TypeScript hack. So just went ahead and committed those minor changes.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Moot now with keyboard controls removed.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Vectors seem mostly appropriately used; resolving.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: The current docs seem to describe this OK, even if it is still just slightly tricky to follow exactly what's going on in glyph generation.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Doesn't seem to work; if I make a growth function that returns all negative values (like multiplying the default one by -1), I get a completely black canvas. Please could you do the abs-ing claimed here and document this aspect? If you don't get to it before I get back from my morning bike ride, I will do it when I get back, no problem.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: OK, resolving.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Moot since key interaction has been removed. Resolving.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Seems OK now.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Overall documentation is good.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Again, moot because keyboard interaction has been removed. Resolving.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Agreed.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Moot now that keyboard interaction has been removed.

**Comment by gwhitney** [comment]
OK, all seems well except for negative growth function values, and I have to check a couple errors I left unresolved above but don't have time this second (they  are probably fine though). Once that's all done I will merge.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Agreed.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: OK, going from n+2 to n+3 doesn't just shift left, it also changes colors because the primes are colored in the order they occur in the sequence, not in numerical order. But I think that is intentional, and I think it is working. Resolving.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: Yes, seems to take the primes that occur as factors, not just those that occur as elements of the sequence.

File: src/visualizers/SeqColor.ts
Comment (gwhitney) [reply src/visualizers/SeqColor.ts]: OK, at this point A095102 is producing a nice rainbowy sequence of glyphs and all 64 appear with the default settings, so closing.

**Comment by gwhitney** [comment]
OK, have tested all of the other open points and resolved them, so I will now take care of the negative growth function values.

**Comment by gwhitney** [comment]
Alright, now just the absolute value of the brightness function is used (and I dropped the outer `abs` from the default brightness function). I don't see anything else, so merging.

**Pull request merged**
